In [3]:
import json
import numpy as np
import tensorflow as tf

In [4]:
# Opening JSON file
f = open('/Users/adityanaik/Desktop/Summer project/medical-question-answer-data-master/icliniqQAs.json')
  
# returns JSON object as 
# a dictionary
data = json.load(f)

# Closing file
f.close()

#### Displaying one question and answer from the dataset

In [5]:
data[0]['answer']

'from your description it appears that you may have anterior knee pain which sometimes presents as pain at the back of the knee. the second possibility is that you have over done your exercise and hamstrings are sore and the lower end of the knee cap is inflamed. either way you should rest and ice the area of pain and give it time. i suggest you add nsaids (nonsteroidal anti-inflammatory drugs) for a week for an additional anti-inflammatory action. for further information consult an orthopaedician and traumatologist online --> <link>'

In [6]:
data[0]['question']

'is it fine to exercise with knee pain?'

##### Preprocessing
Removing the stop words and unnecessary punctuations to extract only useful unique words.

In [7]:
from nltk.corpus import stopwords
from collections import defaultdict

stop = set(stopwords.words('english'))


In [8]:
def CreateCorpus(text):
    corpus = []
    
    for x in text.split():
        corpus.append(x)
    return corpus



In [9]:
test_words = []
for i in range(len(data)):
    test_words1 = CreateCorpus(data[i]['answer'])
    test_words2 = CreateCorpus(data[i]['question'])
    test_words = test_words + test_words1 + test_words2

In [10]:
len(test_words)

72517

In [11]:
test_words

['from',
 'your',
 'description',
 'it',
 'appears',
 'that',
 'you',
 'may',
 'have',
 'anterior',
 'knee',
 'pain',
 'which',
 'sometimes',
 'presents',
 'as',
 'pain',
 'at',
 'the',
 'back',
 'of',
 'the',
 'knee.',
 'the',
 'second',
 'possibility',
 'is',
 'that',
 'you',
 'have',
 'over',
 'done',
 'your',
 'exercise',
 'and',
 'hamstrings',
 'are',
 'sore',
 'and',
 'the',
 'lower',
 'end',
 'of',
 'the',
 'knee',
 'cap',
 'is',
 'inflamed.',
 'either',
 'way',
 'you',
 'should',
 'rest',
 'and',
 'ice',
 'the',
 'area',
 'of',
 'pain',
 'and',
 'give',
 'it',
 'time.',
 'i',
 'suggest',
 'you',
 'add',
 'nsaids',
 '(nonsteroidal',
 'anti-inflammatory',
 'drugs)',
 'for',
 'a',
 'week',
 'for',
 'an',
 'additional',
 'anti-inflammatory',
 'action.',
 'for',
 'further',
 'information',
 'consult',
 'an',
 'orthopaedician',
 'and',
 'traumatologist',
 'online',
 '-->',
 '<link>',
 'is',
 'it',
 'fine',
 'to',
 'exercise',
 'with',
 'knee',
 'pain?',
 'depression',
 'anxiety',
 'r

##### Removing punctuation marks

In [12]:
# Removing punctuations
import string
def RemovePunct(text):
    table = str.maketrans('', '', string.punctuation)
    return text.translate(table)

In [13]:
l = []
for word in test_words:
    l.append(RemovePunct(word))
test_words = l

##### Removing stopwords

In [14]:
def RemoveStopwords(word_list):
    a = []
    for word in word_list:
        if word not in stop:
            a.append(word)
    return a

In [15]:
b = RemoveStopwords(test_words)
test_words = b

In [16]:
# Removing duplicates
c = set(test_words)
test_words = c

In [17]:
len(test_words)

6233

##### Total number of unique words from the iCliniQA dataset is 6233

## Test 1 Embedding generation ability
This test involves checking the ability fo the model to generate embedding for words. The model gets one point if it can and doesnt get a point if it cannot geneate an embedding. The accuracy is then measured by the ratio of the total number of points and the total number of words

### Evaluating GloVe

In [47]:
# Loading the Glove file
import pandas as pd
import csv
words = pd.read_table('/Users/adityanaik/Desktop/Summer project/CODE files/glove.840B.300d.txt', sep=" ", index_col=0, header=None, quoting=csv.QUOTE_NONE)

In [30]:
pos = 0
neg = 0
neg_words = []
for word in test_words:
    try:
        words.loc[word]
        pos += 1
    except:
        neg_words.append(word)
        neg +=1

In [31]:
print('Number of words detected by GloVe', pos)
print('Number of words not detected by Glove ', neg)

Number of words detected by GloVe 5796
Number of words not detected by Glove  437


In [32]:
neg_words

['',
 'doneecg',
 'lowcalorie',
 'donelft',
 'causessurgical',
 'clothesjeans',
 'budecort',
 'cotaryl',
 'ovasafe',
 'agerelated',
 'lenolenic',
 'planwait',
 'rasagandhi',
 'mymox',
 'antipigmentation',
 'duoluton',
 'zungs',
 'decdan',
 'causesanxiety',
 'trapicmf',
 'plansildenafil',
 'neuropraxia',
 'sightproximityodor',
 'foulsmelling',
 'xlinked',
 'reckeweg',
 'ornidazole',
 'monteklc',
 'cottononly',
 'urti',
 'postssri',
 'medroxy',
 'bichromicum',
 'lavanabhaskar',
 'lehyam',
 'diagnosisspace',
 'shuddhi',
 'flutifloft',
 'tomatobased',
 'xaria',
 'causesphysical',
 'hydrozole',
 'levosulpiride',
 'darolacibs',
 'nexpro',
 'godanti',
 'mirap',
 'creactive',
 'causesfluoxetine',
 'transaminitis',
 'femilon',
 'koshta',
 'diagnosisautoimmune',
 'elbowsknees',
 'depiwhite',
 'bilwadi',
 'folvite',
 'hollowbore',
 'glutamicoxaloacetic',
 'kuligai',
 'systane',
 'hypochondroplasia',
 'pegred',
 'neodymiumdoped',
 'flunarin',
 'emefilm',
 'causesthe',
 'deviry',
 'clonapin',
 'lgl

##### Accuracy of the GloVe model is 92%

### Evaluating Word2Vec

In [33]:
from gensim.models import Word2Vec, KeyedVectors
import csv

In [34]:
model = KeyedVectors.load_word2vec_format('/Users/adityanaik/Desktop/Summer project/CODE files/GoogleNews-vectors-negative300.bin', binary = True, limit = 100000)

In [35]:
pos = 0
neg = 0
neg_words = []
for word in test_words:
    try:
        model[word]
        pos += 1
    except:
        neg_words.append(word)
        neg +=1

In [36]:
print('Number of words detected by Word2Vec', pos)
print('Number of words not detected by Word2Vec ', neg)

Number of words detected by Word2Vec 4377
Number of words not detected by Word2Vec  1856


In [37]:
neg_words

['',
 'doneecg',
 'ekgs',
 'fcd',
 'tmt',
 'lowcalorie',
 'splenomegaly',
 'ckd',
 'ethinyl',
 '5ml',
 'rehydrate',
 'donelft',
 'reemergence',
 'celexa',
 'rheumatoid',
 'uroflowmetry',
 'acyclovir',
 'causessurgical',
 'clothesjeans',
 'budecort',
 'cotaryl',
 'transiently',
 'pasteurized',
 'ovasafe',
 'pyuria',
 'agerelated',
 'satiety',
 'hyperthyroidism',
 'specificities',
 'dilation',
 'stabiliser',
 'midline',
 'overthink',
 'yolks',
 'cardioprotective',
 'manometry',
 'celiac',
 'sedimentation',
 'reductase',
 'famotidine',
 'proteolytic',
 'lenolenic',
 'planwait',
 'parenchyma',
 'avr',
 'abrasiveness',
 'rasagandhi',
 'aloes',
 'mymox',
 'foreplay',
 'antipigmentation',
 'hbsag',
 'endothelium',
 'duoluton',
 'intermenstrual',
 '26',
 'dexamethasone',
 'bacillus',
 'zungs',
 'decdan',
 'gris',
 'causesanxiety',
 'glycolic',
 'neurotropic',
 'trapicmf',
 'testes',
 'lifethreatening',
 'esophageal',
 'chalky',
 '40mg',
 'angiogram',
 'analgesics',
 'plansildenafil',
 'neuropr

##### Accuracy of the Word2Vec model is 70%

### Evaluating Fasttext

In [1]:
from gensim.models.fasttext import load_facebook_model

wv = load_facebook_model('/Users/adityanaik/Desktop/Summer project/CODE files/crawl-300d-2M-subword/crawl-300d-2M-subword.bin')

In [18]:
pos = 0
neg = 0
neg_words = []
for word in test_words:
    try:
        wv.wv[word]
        pos += 1
    except:
        neg_words.append(word)
        neg +=1

In [19]:
print('Number of words detected by fasttext', pos)
print('Number of words not detected by fasttext', neg)

Number of words detected by fasttext 6233
Number of words not detected by fasttext 0


In [20]:
neg_words

[]

##### Accuracy of the fasttext model is 100%

#### Measuring the performance on the symptom corpus extracted from the symptom-disease prediction dataset

In [61]:
df = pd.read_csv(r'/Users/adityanaik/Desktop/Summer project/healthcare-chatbot-master/Data/Training.csv')

symptom_corpus = list(df.columns)
symptom_corpus = [x.replace('_', ' ') for x in symptom_corpus]
symptom_corpus = [x.translate(str.maketrans('', '', string.punctuation)) for x in symptom_corpus]
symptom_corpus = [x.translate(str.maketrans('', '', string.digits)) for x in symptom_corpus]
training_df = df
training_df.columns = symptom_corpus
symptom_corpus.remove('prognosis')



In [62]:
symptom_corpus

['itching',
 'skin rash',
 'nodal skin eruptions',
 'continuous sneezing',
 'shivering',
 'chills',
 'joint pain',
 'stomach pain',
 'acidity',
 'ulcers on tongue',
 'muscle wasting',
 'vomiting',
 'burning micturition',
 'spotting  urination',
 'fatigue',
 'weight gain',
 'anxiety',
 'cold hands and feets',
 'mood swings',
 'weight loss',
 'restlessness',
 'lethargy',
 'patches in throat',
 'irregular sugar level',
 'cough',
 'high fever',
 'sunken eyes',
 'breathlessness',
 'sweating',
 'dehydration',
 'indigestion',
 'headache',
 'yellowish skin',
 'dark urine',
 'nausea',
 'loss of appetite',
 'pain behind the eyes',
 'back pain',
 'constipation',
 'abdominal pain',
 'diarrhoea',
 'mild fever',
 'yellow urine',
 'yellowing of eyes',
 'acute liver failure',
 'fluid overload',
 'swelling of stomach',
 'swelled lymph nodes',
 'malaise',
 'blurred and distorted vision',
 'phlegm',
 'throat irritation',
 'redness of eyes',
 'sinus pressure',
 'runny nose',
 'congestion',
 'chest pain',


In [68]:
def GloVeEmbeddings(symptom_word, words):
    # List containing all the embeddings for each word in the sentence
    emb_arr = []
    # Representing the embeddings by taking the mean of each in case its
    for word in symptom_word.split():
        emb_arr.append(words.loc[word].to_numpy())
    emb_arr = np.array(emb_arr)
    fin_emb = np.mean(emb_arr, axis=0)
    return fin_emb

def FasttextEmbeddings(symptom_word, wv):
    # List containing all the embeddings for each word in the sentence
    emb_arr = []
    # Representing the embeddings by taking the mean of each in case its
    for word in symptom_word.split():
        emb_arr.append(wv.wv[word])
    emb_arr = np.array(emb_arr)
    fin_emb = np.mean(emb_arr, axis = 0)
    return fin_emb

def W2VEmbeddings(symptom_word):
    # List containing all the embeddings for each word in the sentence
    emb_arr = []
    # Representing the embeddings by taking the mean of each in case its
    for word in symptom_word.split():
        emb_arr.append(model[word])
    emb_arr = np.array(emb_arr)
    fin_emb = np.mean(emb_arr, axis = 0)
    return fin_emb

def similarityCheck(emb1, emb2):
    # Normalising both vectors
    normalize_emb1 = tf.nn.l2_normalize(emb1, 0)
    normalize_emb2 = tf.nn.l2_normalize(emb2, 0)
    # Calculating cos similarity
    cos_similarity = tf.reduce_sum(tf.multiply(normalize_emb1, normalize_emb2))
    return cos_similarity

### Evaluating the GloVe model for our symptom corpus

In [74]:
pos = 0
neg = 0
neg_words = []
for word in symptom_corpus:
    try:
        GloVeEmbeddings(word)
        pos += 1
    except:
        neg_words.append(word)
        neg +=1

In [75]:
pos

130

In [76]:
neg

2

##### GloVe model can generate embeddings for 130 out of 132 symptom words

### Evaluating W2V model for our symptom corpus

In [82]:
pos = 0
neg = 0
neg_words = []
for word in symptom_corpus:
    try:
        W2VEmbeddings(word)
        pos += 1
    except:
        neg_words.append(word)
        neg +=1

In [83]:
pos

95

In [84]:
neg

37

##### W2V model can only generate embedding for 95 our of 132 symptoms

### Evaluating the Fasttext model for our symptom corpus

In [90]:
pos = 0
neg = 0
neg_words = []
for word in symptom_corpus:
    try:
        FasttextEmbeddings(word)
        pos += 1
    except:
        neg_words.append(word)
        neg +=1

In [91]:
pos

132

In [92]:
neg

0

##### Fasttext model can generate embeddings for all symptom words

## Test 2 synonym accuracy

In [4]:
import pandas as pd

syn = pd.read_csv(r'/Users/adityanaik/Desktop/archive/synonyms.csv')
syn.head()

lemma part_of_speech                             synonyms
0  .22-caliber      adjective  .22 caliber;.22 calibre;.22-calibre
1  .22-calibre      adjective  .22 caliber;.22-caliber;.22 calibre
2  .22 caliber      adjective  .22-caliber;.22 calibre;.22-calibre
3  .22 calibre      adjective  .22 caliber;.22-caliber;.22-calibre
4  .38-caliber      adjective  .38 caliber;.38 calibre;.38-calibre

In [23]:
a = list(syn['synonyms'][300:])
a

['dormant',
 'detestable;obscene;repugnant;repulsive',
 'bide;abide;stay|digest;endure;stick out;stomach;bear;stand;tolerate;support;brook;abide;suffer;put up',
 'enduring;imperishable',
 'low;low-down;miserable;scummy;scurvy|unhopeful',
 'Abkhaz;Abkhazian;Abkhas;Abkhasian|Abkhaz;Abkhazia|Abkhaz;Abkhas',
 'Abkhaz;Abkhazian',
 'Abkhaz;Abkhazian;Abkhas;Abkhasian|Abkhazian;Abkhasian',
 'Abkhaz;Abkhazian',
 'ablate|ablate',
 'ablative case',
 'aflame;aroused|afire;aflame;aflare;alight;on fire|inflamed;reddened',
 'capable|able-bodied',
 'able',
 'efflorescent',
 'cleansing',
 'Abnaki;Abenaki|Abnaki',
 'Abnaki',
 'unnatural',
 'detestable;execrable;odious|atrocious;awful;dreadful;painful;terrible;unspeakable',
 'Aborigine;Abo;Aboriginal;native Australian;Australian Aborigine|native;indigen;indigene;aborigine',
 'Aboriginal|native',
 'primal;primeval;primaeval;primordial',
 'aborticide;abortion-inducing drug',
 'stillborn;unsuccessful',
 'abulic',
 'abound|abound;burst;bristle',
 'galore',
 

In [35]:
b = []
c = []
for words in a:
    try:
        if len(words.split(";")) > 1:
            try:
                b.append(words.split(";")[0])
                c.append(words.split(";")[1])
            except:
                pass
    except:
        pass
    


##### Selecting 1000 words and their synonyms from the dataset

In [41]:
main1 = b[0:1000]
main2 = c[0:1000]

### Evaluating the GloVe model for matching synonyms

In [57]:

result = []
for i in range(len(main1)):
        if len(main1[i].split())==1 or len(main2[i].split())==1:
            try:
                emb1 = GloVeEmbeddings(main1[i], words).astype('float32')
                emb2 = GloVeEmbeddings(main2[i], words).astype('float32')
                cos = similarityCheck(emb1, emb2)
                if cos > 0.65:
                    result.append(1)
                else:
                    result.append(0)
            except:
                pass

            
            

In [59]:
len(result)

672

In [60]:
np.sum(result)

156

The GLoVe model could only generate embeddings for 672 words among the 1000 words out of which it could match only 156 words to its synonyms giving it an accuracy of 23%

### Evaluating the Fasttext model for matching synonyms

In [75]:
result = []
for i in range(len(main1)):
        if len(main1[i].split())==1 or len(main2[i].split())==1:
            try:
                emb1 = FasttextEmbeddings(main1[i], wv).astype('float32')
                emb2 = FasttextEmbeddings(main2[i], wv).astype('float32')
                cos = similarityCheck(emb1, emb2)
                if cos > 0.65:
                    result.append(1)
                else:
                    result.append(0)
            except:
                pass

In [76]:
len(result)

973

In [77]:
np.sum(result)

293

The fasttext model could generate embeddings for 973 words out of 1000 out of which it could match the synonyms of 293 words giving it an accuracy of 30%

##### Based on the results of test 1 and test 2 we select the fasttext model to be used in our final implementation.